src/foam/nanos/auth/PasswordHashingDAO.java

```diff
@@ -0,0 +1,32 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.util.Password;
+
+public class PasswordHashingDAO
+    extends ProxyDAO
+{
+  public PasswordHashingDAO(X x, DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    PropertyInfo prop = (PropertyInfo) obj.getClassInfo().getAxiomByName("password");
+    if ( prop != null && prop.get(obj) instanceof String ) {
+      prop.set(obj, Password.hash((String) prop.get(obj)));
```

Comment:
Hashing without salting isn't good because someone can just hash a dictionary of common passwords to get a dictionary of common hashes.

Reply:
Please refer to the Password utility class. A salt is randomly generated and used when hashing.

Reply:
Great!

Reply:
@eatonk @kgrgreer If we're trying to defend against this kind of attack by using a salt but sending the salt to the client, doesn't that defeat the purpose?

See https://github.com/foam-framework/foam2/blob/master/src/foam/util/Password.java#L87 for the return value of `Password.hash()` and https://github.com/foam-framework/foam2/blob/master/src/foam/nanos/auth/UserPasswordHashingDAO.java#L32 for where we set the `password` property on the User model to the result of a call to `Password.hash()`. That means we're sending the salt and the hash of the password to the user on the client, which would be just as (in)secure as sending just the password hash and not using a salt at all, right?

Shouldn't we be storing the salt somewhere but never sending it to the client?